slither/detectors/statements/unused_imports.py

```diff
@@ -0,0 +1,543 @@
+from logging import Logger
+from pathlib import PurePath
+from typing import Dict, Set, TYPE_CHECKING, List
+
+from crytic_compile.utils.naming import Filename
+
+from slither.core.compilation_unit import SlitherCompilationUnit
+from slither.core.declarations import Contract
+from slither.core.scope.scope import FileScope
+from slither.core.solidity_types import TypeAliasTopLevel, TypeAliasContract
+from slither.core.source_mapping.source_mapping import SourceMapping
+from slither.core.variables.variable import Variable
+from slither.detectors.abstract_detector import AbstractDetector, DetectorClassification, DETECTOR_INFO
+from slither.utils.output import Output
+
+if TYPE_CHECKING:
+    from slither import Slither
+
+
+class UnusedImports(AbstractDetector):
+    """
+    Unused imports detector
+    """
+
+    ARGUMENT = "unused-imports"
+    HELP = "Detect unused imports"
+    IMPACT = DetectorClassification.INFORMATIONAL
+    CONFIDENCE = DetectorClassification.HIGH
+
+    WIKI = "https://github.com/crytic/slither/wiki/Detector-Documentation#unused-imports"
+
+    WIKI_TITLE = "Unused imports"
+    WIKI_DESCRIPTION = (
+        "Detects imports that are not used by source files. Currently, files with cyclic imports and files with "
+        "'import {...} from' directives are not supported."
+    )
+    WIKI_RECOMMENDATION = "Remove unused imports"
+
+    analysed_files: Set[str] = set()
```

Comment:
Using class variables for things that are going to be dynamically change by the objects seem really error prone. Can you explain the reasoning here?

Reply:
Sure, my reasoning about this was to store some data globally so that it persists through the detector's execution. 

If my understanding is correct (please correct me if I'm wrong), detector will be run on each file individually. However, the detector analyses all the files in the compilation unit and prints recommendations for all of them. So, the `analysed_files` set tells the detector not to print findings for files that were already analysed if they belong to many compilation units.

The reasoning behind the second class level variable, `import_containers`, was to store files that only aggregate imports, so that the detector doesn't have to check if they are "import containers" all over again. However, I forgot to add the check in `_is_import_container`, so it doesn't provide any optimization at the moment and when I think about it right now, such files don't contain too much code, so even if they are analysed once again, it wouldn't bring too much overhead - I should probably change it to instance variable.

Reply:
It makes perfect sense. Thank you for the explanation. I have changed `import_containers` to an instance variable and removed `analysed_files` entirely. 

I was running `slither . --detect unused-imports` in the test directory and it was reporting each file individually and hence my confusion. But when I tried it on a hardhat project, it doesn't duplicate outputs.